torch_em/trainer/default_trainer.py

```diff
@@ -76,64 +79,119 @@ def iteration(self):
     def epoch(self):
         return self._epoch
 
-    @classmethod
-    def from_checkpoint(cls, checkpoint_folder, name="best", device=None):
-        save_path = os.path.join(checkpoint_folder, f"{name}.pt")
-        if not os.path.exists(save_path):
-            raise ValueError(f"Cannot find checkpoint {save_path}")
-        save_dict = torch.load(save_path, map_location=device)
-
-        init_data = save_dict["init"]
-        model_p, model_m = init_data["model_class"].rsplit(".", 1)
-        model_class = getattr(import_module(model_p), model_m)
-        model = model_class(**init_data["model_kwargs"])
+    class Deserializer:
+        """
+        Determines how to deserialize the trainer kwargs from serialized 'init_data'
+
+        Examples:
+            To extend the initialization process you can inherite from this Deserializer in an inherited Trainer class.
+            Note that `DefaultTrainer.Deserializer.load_generic()` covers most cases already.
+
+            This example adds `the_answer` kwarg, which requires 'calculations' upon initialization
+            >>> class MyTrainer(DefaultTrainer):
+            >>>     def __init__(self, *args, the_answer: int, **kwargs):
+            >>>         super().__init__(*args, **kwargs)
+            >>>         self.the_answer = the_answer
+            >>>
+            >>>     class Deserializer(DefaultTrainer.Deserializer):
+            >>>         def __call__(self, name, *args, **kwargs):
+            >>>             if name == "the_answer":
+            >>>                 return self.load_the_answer()
+            >>>             else:
+            >>>                 return super().__call__(name, *args, **kwargs)
+            >>>
+            >>>         def load_the_answer(self):
+            >>>             if self.device.type == "cpu":
+            >>>                 return ...  # complex 'calculation' to arrive at the answer
+            >>>             else:
+            >>>                 return 42
+        """
+
+        def __init__(self, init_data: dict, save_path: str, device: Union[str, torch.device]):
+            self.init_data = init_data
+            self.save_path = save_path
+            self.device = torch.device(self.init_data["device"]) if device is None else torch.device(device)
+
+        def __call__(
+            self,
+            name: str,
+            *dynamic_args,
+            optional=False,
+            only_class=False,
+            dynamic_kwargs: Optional[Dict[str, Any]] = None,
```

Comment:
Is there a particular reason you're not using `**dynamic_kwargs` here?

Reply:
yes, `**dynamic_kwargs` would not allow to `optional` (or `only_class`) as a kwarg

Reply:
Maybe I am missing something, but I don't see why this is the case. As far as I can see this is perfectly valid python:
```python
def my_func(
    kwarg_name, *args,
    optional=False, only_class=False,
    **kwargs
):
    print(optional)

my_func("x", "y", optional=True, blub="xyz")
```
(prints True)